{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2017-05-15 : running the main openRetina scripts\n",
    "---\n",
    "\n",
    "Progress of my work : After some weeks of reverse-engineering the OpenRetina scripts, I understood them enough to work direclty on them. L. Perrinet and I built some testing scripts that allowed me to try these scripts. The first one, with a fully-noised input, was run successfully. I still have some problems to run the second one, that use a local camera as an input.\n",
    "\n",
    "- [ ] Run the local testing script\n",
    "    - [ ] Solve the empty capture problem\n",
    "    \n",
    "Since the code() function seems to receive empty inputs, the problem may come from the run() function.  \n",
    "Trying to find the source of that problem.  \n",
    "\n",
    "    def run(self):\n",
    "        (...)\n",
    "        print('dbg_key1', self.frame)\n",
    "        data = self.code(self.frame)\n",
    "        (...)\n",
    "        \n",
    "Output : \n",
    "\n",
    "    dbg_key1 None\n",
    "    \n",
    "Then :\n",
    "\n",
    "    def grab(self):\n",
    "        (...)\n",
    "        print('dbg_key1', frame)\n",
    "        print('dbg_key2', self.cap.read())\n",
    "        (...)\n",
    "        \n",
    "Output : \n",
    "    \n",
    "    dbg_key1 None\n",
    "    dbg_key2 (False, None)\n",
    "    \n",
    "The problem seem to come from the PhotoReceptor() class, as trying to call it alone don't work : \n",
    "\n",
    "    from openRetina import openRetina\n",
    "    from openRetina import PhotoReceptor\n",
    "    import cv2\n",
    "\n",
    "    cam = PhotoReceptor(640,480)\n",
    "    cam.grab()\n",
    "    \n",
    "Output : \n",
    "\n",
    "    Unable to capture video\n",
    "    dbg_key1 (False, None)\n",
    "\n",
    "I had some internet problems that slowed down my work. My computer get randomly disconnected, forcing me to restart it to make it work again.  \n",
    "The script problem come from the PhotoReceptor class, as : \n",
    "\n",
    "    (...)\n",
    "    if not self.cap.isOpened(): \n",
    "        print('Camera is not opened')\n",
    "        stop\n",
    "    print('dbgkey1')\n",
    "    (...)\n",
    "    \n",
    "Output :\n",
    "\n",
    "    Camera is not opened\n",
    "    \n",
    "I'm missing something. While my own *camera.py* and the official *openRetina* PhotoReceptor classes are fairly similar, the openRetina one don't work. \n",
    "\n",
    "    Before a downscale of 1 dim1 : 480, dim2 : 640\n",
    "    Using OpenCV\n",
    "    After a downscale of 1, dim1 : 0.0, dim2 : 0.0\n",
    "    \n",
    "The null dimensions surely come from the closed camera. Why is the camera not opening?  \n",
    "\n",
    "- [ ] Run the local testing script\n",
    "    - [ ] Solve the non-opening camera problem\n",
    "\n",
    "OpenCV documentation suggested to use the following function to force it, but it doesn't change anything (see : http://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_gui/py_video_display/py_video_display.html#capture-video-from-camera) :\n",
    "\n",
    "    self.cap.open(0)\n",
    "    \n",
    "When used through Cheese (capturing program), the camera open and record just fine.  \n",
    "This command line give informations about the devices connected to the computer's USB bus, including the camera (first one to appear, see : https://help.ubuntu.com/community/Webcam) :\n",
    "\n",
    "    lsusb -v\n",
    "    \n",
    "Even when using a simple script that I already used and that was working, the camera don't launch and no error is given :\n",
    "\n",
    "    import cv2\n",
    "\n",
    "    cap = cv2.VideoCapture(0)\n",
    "    if not cap.isOpened():\n",
    "        print('Cam not opnd')\n",
    "    while(True):\n",
    "        ret,frame = cap.read()\n",
    "        cv2.imshow('frame', frame)\n",
    "        if cv2.waitKey(1) & 0xFF == ord('q'):\n",
    "            break\n",
    "    cap.release()\n",
    "    cv2.destroyAllWindows()\n",
    "\n",
    "I'm using the last openCV version (3.2.0.7), maybe that's what is causing that bug. But the VideoCapture is still called the same way than the previous versions I used (API of the function for that version : http://docs.opencv.org/3.0-beta/modules/videoio/doc/reading_and_writing_video.html?highlight=videocapture#videocapture-videocapture).  \n",
    "    The fact that the camera isn't opening on my personnal computer that use the same openCV version reinforce the idea that it is linked to the openCV version used (as suggested here : http://answers.opencv.org/question/88857/videocapture-is-not-working-in-opencv-300-or-the-310-in-windows-7/ and suggested by lot of recent questions on internet about the same issue, eg.: http://answers.opencv.org/question/136678/opencv-error-assertion-failed-sizewidth0-sizeheight0-in-cvimshow-python/, https://github.com/opencv/opencv/issues/8471 and https://github.com/opencv/opencv/issues/8358).  \n",
    "I need to ask L. Perrinet which version is installed on his computer, since the scripts worked on it.  \n",
    "All these links suggest that the issue is caused by the version 3.2.x, so I tried to downgrade to 3.1.x, the previous version :\n",
    "\n",
    "    pip3 uninstall opencv-python\n",
    "    pip install opencv-python==3.1.0 --user\n",
    "    \n",
    "But it didn't solved the issue.  \n",
    "I'm working on that issue since this morning and I'm still not able to solve it.  I'll talk with L. Perrinet about that.\n",
    "\n",
    "In the meantime, I tried to run *2017-05-15_using_openretina* and *2017-05-15_using_openretina_server*.\n",
    "\n",
    "- [X] Run the openRetina script\n",
    "\n",
    "It works! A window is opening and show several frames captured, with a red/blue color coding on a black background, as programmed by the openRetina script.  \n",
    "Still, several non-critical errors are showing, that may decrease the script performance.\n",
    "\n",
    "    *** Pi side ***\n",
    "    \n",
    "    mmal: mmal_vc_port_enable: failed to enable port vc.null_sink:in:0(OPQV): ENOSPC\n",
    "    mmal: mmal_port_enable: failed to enable connected port (vc.null_sink:in:0(OPQV))0x16dcd70 (ENOSPC)\n",
    "    mmal: mmal_connection_enable: output port couldn't be enabled\n",
    "    \n",
    "    ðŸ’€  Could not load visualisation\n",
    "    \n",
    "    Traceback (most recent call last):\n",
    "      File \"2017-05-15_using_openretina.py\", line 13, in <module>\n",
    "        output=['stream'] # output: can be stream, display, ...\n",
    "      File \"/home/pi/openRetina/src/openRetina.py\", line 146, in __init__\n",
    "        self.camera = PhotoReceptor(self.w, self.h)\n",
    "      File \"/home/pi/openRetina/src/openRetina.py\", line 46, in __init__\n",
    "        self.cap = picamera.PiCamera(cam_id)\n",
    "      File \"/usr/lib/python3/dist-packages/picamera/camera.py\", line 433, in __init__\n",
    "        self._init_preview()\n",
    "      File \"/usr/lib/python3/dist-packages/picamera/camera.py\", line 513, in _init_preview\n",
    "        self, self._camera.outputs[self.CAMERA_PREVIEW_PORT])\n",
    "      File \"/usr/lib/python3/dist-packages/picamera/renderers.py\", line 558, in __init__\n",
    "        self.renderer.inputs[0].connect(source).enable()\n",
    "      File \"/usr/lib/python3/dist-packages/picamera/mmalobj.py\", line 2212, in enable\n",
    "        prefix=\"Failed to enable connection\")\n",
    "      File \"/usr/lib/python3/dist-packages/picamera/exc.py\", line 184, in mmal_check\n",
    "        raise PiCameraMMALError(status, prefix)\n",
    "    picamera.exc.PiCameraMMALError: Failed to enable connection: Out of resources\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2017-05-16 : capturing a video file with the Pi and sending it to a server\n",
    "---\n",
    "\n",
    "These errors seem to be linked to picamera's mmalobj module (complete API : http://picamera.readthedocs.io/en/latest/api_mmalobj.html). They don't show at every run.  \n",
    "\n",
    "Now that the openRetina script is correctly running, I can begin to apply modifications to it. In accordance with L. Perrinet, I'll insert a simple script to capture and save a video file and then modify the script to allow it to work with that file, instead of a stream. That'll allow to bypass the networking limitations, like bandwidth consumption.  \n",
    "\n",
    "I'll then capture a video with a stable background and a moving object (eg. a hand), so some of the pixels will change value while other will keep the same over time. The objective is to modify the program to only code (send to server) information about the changing pixels instead of coding the luminance contrast as it does now.\n",
    "\n",
    "About the poster that'll act as my final presentation, L. Perrinet thinks that it may be better to show what I've learned (python classes, git and ssh use, raspberry manipulation, biomimetism, ...) instead of my results.\n",
    "\n",
    "- [X] Update openRetina script to get more informative prints\n",
    "- [X] Write a simple script to capture a video to a file\n",
    "\n",
    "More informations about the h264 video files format : https://en.wikipedia.org/wiki/H.264/MPEG-4_AVC  \n",
    "At first, making a local (on the Pi) video capture script without any networking to make trials (script inspired by http://picamera.readthedocs.io/en/latest/recipes1.html#recording-video-to-a-file) :\n",
    "\n",
    "    import picamera\n",
    "\n",
    "    picam = picamera.PiCamera\n",
    "    picam.resolution = (640,480)              #Defining the camera resolution\n",
    "    picam.start_recording('video_picam.h264') #Start recording in 'filename.filetype'\n",
    "    picam.wait_recording(10)                  #Better integrated time.wait() that allow to continually check for errors\n",
    "    picam.stop_recording()                    #Stop recording\n",
    "\n",
    "But that script output an error (not present on the internet example, my be induced by the that its the Pi that run the script) : \n",
    "\n",
    "    picam.start_recording('video_picam.h264') #Start recording in 'filename.filetype'\n",
    "    TypeError: start_recording() missing 1 required positional argument: 'output'\n",
    "    \n",
    "After modifying the concerned line a bit, and even if the output argument is the only non-optionnal (see API here : http://picamera.readthedocs.io/en/latest/api_camera.html#picamera.PiCamera.start_recording), the script still asks for more 'missing' arguments : \n",
    "\n",
    "    (...)\n",
    "    picam.start_recording(output='video_picam', format='h264') #Start recording in 'filename.filetype'\n",
    "    (...)\n",
    "\n",
    "Output : \n",
    "\n",
    "    picam.start_recording(output='video_picam', format='h264') #Start recording in 'filename.filetype'\n",
    "    TypeError: start_recording() missing 1 required positional argument: 'self'\n",
    "    \n",
    "That was an human error, and I simply solved the problem by going from :\n",
    "\n",
    "    (...)\n",
    "    picam = picamera.PiCamera\n",
    "    (...)\n",
    "\n",
    "To : \n",
    "\n",
    "    (...)\n",
    "    picam = picamera.PiCamera()\n",
    "    (...)\n",
    "    \n",
    "- [ ] Send the video file to a server\n",
    "\n",
    "Trying to write duo of client/server scripts to send/rcv the video file that was recorded. I do this to try my parameters before blend it in the main openRetina script. By I'm facing a newtorking problem, as the Pi-sided script output :\n",
    "\n",
    "    (...)\n",
    "    socket.bind(\"tcp://*:5555\")\n",
    "    File \"socket.pyx\", line 444, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:4124)\n",
    "    File \"checkrc.pxd\", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:6103)\n",
    "    zmq.error.ZMQError: Address already in use\n",
    "    \n",
    "And the command (more informations here : http://stackoverflow.com/questions/19159771/recovering-from-zmq-error-zmqerror-address-already-in-use) : \n",
    "\n",
    "    netstat -ltnp tcp\n",
    "\n",
    "Output : \n",
    "\n",
    "    (...)\n",
    "    Active Internet connections (only servers)\n",
    "    Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name\n",
    "    tcp        0      0 0.0.0.0:5555            0.0.0.0:*               LISTEN      2006/python3    \n",
    "    tcp        0      0 0.0.0.0:5556            0.0.0.0:*               LISTEN      1849/python3    \n",
    "    (...)\n",
    "\n",
    "A connection can be manually closed with the command : \n",
    "\n",
    "    kill -9 <pid>\n",
    "\n",
    "That issue can be resolved by adding encapsulation (try/except/finally) loops in the script.\n",
    "\n",
    "Since I already did it, I first tried to build a network using the REQ/REP zmq's parameters, but other options can be tried : http://zguide.zeromq.org/py:fileio1, http://zguide.zeromq.org/py:fileio3, http://zguide.zeromq.org/py:fileio3  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2017-05-17 : correctly encoding videos to send them to the server\n",
    "---\n",
    "I still have problems with my internet connection around 9, that makes me restart the Mac computer.  \n",
    "\n",
    "The Pi/computer connection is correctly initiated, but the file is not send, as that error message is outputed :\n",
    "\n",
    "    File \"2017-05-16_capture_videos_with_openretina_client.py\", line 17, in <module>\n",
    "        socket.send(Vid)\n",
    "      File \"socket.pyx\", line 574, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:5390)\n",
    "      File \"socket.pyx\", line 621, in zmq.backend.cython.socket.Socket.send (zmq/backend/cython/socket.c:5157)\n",
    "      File \"socket.pyx\", line 168, in zmq.backend.cython.socket._send_copy (zmq/backend/cython/socket.c:2021)\n",
    "      File \"buffers.pxd\", line 200, in buffers.asbuffer_r (zmq/backend/cython/socket.c:7091)\n",
    "      File \"buffers.pxd\", line 151, in buffers.asbuffer (zmq/backend/cython/socket.c:6523)\n",
    "    TypeError: <_io.TextIOWrapper name='video_picam' mode='r' encoding='UTF-8'> does not provide a buffer interface.\n",
    "\n",
    "More informations about transferring files with zmq : http://zguide.zeromq.org/php:chapter7#Transferring-Files  \n",
    "(More informations about the raise statement : https://docs.python.org/3/tutorial/errors.html#raising-exceptions)  \n",
    "(Some informations about the assert statement, equivalent of a \"if not\" line : http://stackoverflow.com/questions/5142418/what-is-the-use-of-assert-in-python)  \n",
    "\n",
    "Trying to insert the zmq simplest method to send files (see : http://zguide.zeromq.org/py:fileio1) into the *_capture_videos_with_openretina* scripts.  \n",
    "\n",
    "After insetring it and correcting some human-made breaking errors, I face that error :\n",
    "\n",
    "    data = file.read(CHUNK_SIZE)\n",
    "    File \"/usr/lib/python3.4/codecs.py\", line 313, in decode\n",
    "    (result, consumed) = self._buffer_decode(data, self.errors, final)\n",
    "    UnicodeDecodeError: 'utf-8' codec can't decode byte 0xac in position 8: invalid start byte\n",
    "    \n",
    "- [ ] Send the video file to a server\n",
    "    - [X] Solve the encoding format problem\n",
    "\n",
    "So there seems it is a data encoding problem, but it looks weird that the error is only triggered at the 8th position.  \n",
    "I'll try to change the recording format, decided at :\n",
    "\n",
    "    (...)\n",
    "    cam.start_recording(output='video_picam2', format='h264')\n",
    "    (...)\n",
    "    \n",
    "When switching format to rgb, and even without changing the recording time, the camera stayed opened for a longer time than usual. Maybe the rgb format -a raw 24-bit format- is heavier so the program is severly slowed down? Anyway the error stayed.  \n",
    "\n",
    "I tried to explicilty tell the encoding format when opening the video file (as suggested here : http://stackoverflow.com/questions/19699367/unicodedecodeerror-utf-8-codec-cant-decode-byte) in that line :\n",
    "\n",
    "    (...)\n",
    "    file = open(\"video_picam\",\"r\", encoding=\"rgb\")\n",
    "    (...)\n",
    "        \n",
    "The script now output that error : \n",
    "\n",
    "    (...)\n",
    "    file = open(\"video_picam\",\"r\", encoding=\"rgb\")\n",
    "    LookupError: unknown encoding: rgb\n",
    "    \n",
    "I modified again that line to make the script open the file as bytes (more informations about open() : https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files) :\n",
    "\n",
    "    (...)\n",
    "    file = open(\"video_picam\",\"rb\", encoding=\"rgb\")\n",
    "    (...)\n",
    "    \n",
    "The error changed again, asking me to reverse one of my previous actions : \n",
    "\n",
    "    (...)\n",
    "    file = open(\"video_picam\",\"rb\", encoding=\"rgb\")\n",
    "    ValueError: binary mode doesn't take an encoding argument\n",
    "\n",
    "I've almost reached my goal. The computer-sided script output : \n",
    "\n",
    "    Connected\n",
    "    Sending request\n",
    "    3 chuncks received, 418001 bytes\n",
    "    Connection closed\n",
    " \n",
    "But also an error, easily corrected : \n",
    "\n",
    "    (...)\n",
    "    NameError: name 'pipe' is not defined\n",
    "    \n",
    "The chunks seems correctly transferred, but they're not built into a file so I can't work on them later. I may need to concatenate them into one.  \n",
    "Problem : The Pi-sided script isn't closing after the computer-sided one did.  \n",
    "\n",
    "- [ ] Send the video file to a server\n",
    "    - [X] Solve the non-closing script problem\n",
    "\n",
    "The script was trapped in a infinite *while* loop. Solved it by adding a *break* line at the end of it :\n",
    "\n",
    "    (...)\n",
    "    if not data: break\n",
    "\n",
    "- [ ] Send the video file to a server\n",
    "    - [ ] Rebuild the received file chunks into a file \n",
    "    \n",
    "Note that the script is simple and don't control at all the information flow, so it face an overflow risk for bigger files. Its possible to later modify that script by following the zmq documentation (here : http://zguide.zeromq.org/php:chapter7#Transferring-Files)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2017-05-18\n",
    "---\n",
    "I talked with V. Boutin about my work and he asked me why did I built a script to record a video in a file then send it to the server instead of working with a stream like before to make the server build the video file. He thinks the method I used is much slower and more energy consuming.  \n",
    "I don't agree with that point. Of course my method is slower than using a direct stream, but it allow to have a copy of the video file on both the Pi and the computer so later I may implement a way to compare the pre- and post-processing files, plus it seems easier to do some pre-processing on the video files before sending it to the computer if we have a copy of it in the Pi.  \n",
    "Regarding the energy consumption, I think it is smaller with my method because the video recording and sending are separated process, allowing to record the video only one time and to do some pre-processing/sending the video whenever we want or have free resources on the Pi.\n",
    "\n",
    "Idea for later : it may be interesting to add a way to delete the video file in the Pi, to not overfill its memory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Outline\n",
    "---\n",
    "\n",
    "- [ ] Run the local testing script\n",
    "    - [ ] Solve the empty capture problem\n",
    "    - [ ] Solve the non-opening camera problem\n",
    "- [X] Run the openRetina script\n",
    "- [X] Update openRetina script to get more informative prints\n",
    "- [X] Write a simple script to capture a video to a file\n",
    "- [ ] Send the video file to a server\n",
    "    - [X] Solve the encoding format problem\n",
    "    - [X] Solve the non-closing script problem\n",
    "    - [ ] Rebuild the received file chunks into a file\n",
    "- [ ] Capture a video with a stable background and moving object\n",
    "- [ ] [OPT]  Add a way to delete the video file on the Pi after sending and processing are done."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
